influxdb_store.go

```diff
@@ -0,0 +1,331 @@
+	response, err := in.con.Query(influxDBClient.Query{
+		Command:  command,
+		Database: dbName,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if response.Error() != nil {
+		return nil, response.Error()
+	}
+	// Expecting one result, since a single query is executed.
+	if len(response.Results) != 1 {
+		return nil, errors.New("unexpected number of results for an influxdb single query")
+	}
+	return &response.Results[0], nil
+}
+
+func (in *InfluxDBStore) init(server *influxDBServer.Server) error {
+	in.server = server
+	url, err := url.Parse(fmt.Sprintf("http://%s:%d", influxDBClient.DefaultHost, influxDBClient.DefaultPort))
+	if err != nil {
+		return err
+	}
+	con, err := influxDBClient.NewClient(influxDBClient.Config{URL: *url})
+	if err != nil {
+		return err
+	}
+	in.con = con
+	if err := in.createDBIfNotExists(); err != nil {
+		return err
+	}
+	in.tracesPerPage = defaultTracesPerPage
+	return nil
+}
+
+func (in *InfluxDBStore) removeSpanIfExists(id SpanID) error {
+	cmd := fmt.Sprintf(`
+		DROP SERIES FROM spans WHERE trace_id = '%s' AND span_id = '%s' AND parent_id = '%s'
+	`, id.Trace.String(), id.Span.String(), id.Parent.String())
+	_, err := in.executeOneQuery(cmd)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func annotationsFromRow(r *influxDBModels.Row) (*Annotations, error) {
+	// Actually an influxDBModels.Row represents a single InfluxDB serie.
+	// r.Values[n] is a slice containing span's annotation values.
+	var fields []interface{}
+	if len(r.Values) == 1 {
+		fields = r.Values[0]
+	}
+	// len(r.Values) might be greater than one - meaning there are
+	// some spans to drop, see: InfluxDBStore.Collect(...).
+	// If so last one is picked.
+	if len(r.Values) > 1 {
+		fields = r.Values[len(r.Values)-1]
+	}
+	annotations := make(Annotations, len(fields))
+	// Iterates over fields which represent span's annotation values.
+	for i, field := range fields {
+		// It is safe to do column[0] (eg. 'Server.Request.Method')
+		// matches fields[0] (eg. 'GET')
+		key := r.Columns[i]
+		var value []byte
+		switch field.(type) {
+		case string:
+			value = []byte(field.(string))
+		case nil:
+		default:
+			return nil, fmt.Errorf("unexpected field type: %v", reflect.TypeOf(field))
+		}
+		a := Annotation{
+			Key:   key,
+			Value: value,
+		}
+		annotations = append(annotations, a)
+	}
+	return &annotations, nil
+}
+
+func newSpanFromRow(r *influxDBModels.Row) (*Span, error) {
+	span := &Span{}
+	traceID, err := ParseID(r.Tags["trace_id"])
+	if err != nil {
+		return nil, err
+	}
+	spanID, err := ParseID(r.Tags["span_id"])
+	if err != nil {
+		return nil, err
+	}
+	parentID, err := ParseID(r.Tags["parent_id"])
+	if err != nil {
+		return nil, err
+	}
+	span.ID = SpanID{
+		Trace:  ID(traceID),
+		Span:   ID(spanID),
+		Parent: ID(parentID),
+	}
+	return span, nil
+}
+
+func NewInfluxDBStore(c *influxDBServer.Config, bi *influxDBServer.BuildInfo) (*InfluxDBStore, error) {
+	//TODO: add Authentication.
```

Comment:
related: I wonder if it's possible for us to specify the InfluxDB server config file but as a Go structure instead? I think this would be preffered in general.